File: 2. Structural Design Patterns/Flyweight/Java/FlyweightTemplate.java
// Flyweight Pattern – Java Template

import java.util.*;

class Flyweight {
    // intrinsic, immutable data
}

class FlyweightFactory {
    private static final Map<String, Flyweight> cache = new HashMap<>();
    static Flyweight get(String key){
        return cache.computeIfAbsent(key, k -> new Flyweight());
    }
}

class Context {
    private final Flyweight fw;
    Context(String key){ this.fw = FlyweightFactory.get(key); }
    void operation(/* extrinsic params */){ /* use fw + extrinsic */ }
}

public class FlyweightTemplate {
    public static void main(String[] args){ new Context("key").operation(); }
}

File: 2. Structural Design Patterns/Flyweight/Kotlin/FlyweightTemplate.kt
// Flyweight Pattern – Kotlin Template

class Flyweight { /* intrinsic, immutable data */ }

object FlyweightFactory {
    private val cache = mutableMapOf<String, Flyweight>()
    fun get(key: String): Flyweight = cache.getOrPut(key) { Flyweight() }
}

class Context(key: String) {
    private val fw = FlyweightFactory.get(key)
    fun operation(/* extrinsic params */) { /* use fw + extrinsic */ }
}

fun main(){ Context("key").operation() }
